### PR TITLE
Object graph include function names

### PR DESCRIPTION
**Pull request by jingt06**
- Change Object Graph structure: 
    - functions is a map of form {id: functionName}.
- some useful methods:
    - getFunctionName: Get name of a given function id.
   -  getRoot: Get the root id.
   -  getPropertiesIds: Get the properties and id belong to the given id.

### CODE REVIEW

File: src/ObjectGraph.es6.js [hunk 1]
```diff
@@ -224,7 +224,18 @@ ObjectGraph.prototype.visitObject = function(o) {
...
-  if ( typeof o === 'function' ) this.functions.push(id);
+  if ( typeof o === 'function' ) {
+    if(o.name){
+      this.functions[id] = o.name;
+    } else {
+      // In IE, function does not have name property,
+      // have to use regular expression to find function name.
+      this.functions[id] = o.toString().
+        split(/[\s\(]/g).filter(
+          function(a){return a!==''}
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: nit: spacing `function() { a !== ''; }`.

File: src/ObjectGraph.es6.js [hunk 2]
```diff
@@ -260,7 +271,7 @@ ObjectGraph.prototype.cloneWithout = function(withoutIds) {
   clone.data = _.cloneDeep(this.data);
   clone.metadata = _.cloneDeep(this.metadata);
   clone.protos = _.cloneDeep(this.protos);
-  clone.functions = Array.from(this.functions);
+  clone.functions = _.cloneDeep(this.functions);
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js:276]: Let's put function metadata in `metadata`.

File: src/ObjectGraph.es6.js [hunk 3]
```diff
@@ -410,20 +424,29 @@ ObjectGraph.prototype.getType = function(id) {
...
+};
+
+// Interface method: get the root of object graph.
+ObjectGraph.prototype.getRoot = function() {
+  return this.root;
 };
 
+// Interface method: Get attribute and id belong to given id.
+ObjectGraph.prototype.getPropertiesIds = function(id){
+  return this.data[id]
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: I'm a little anxious about exposing the raw data object for clients to muck with. Perhaps `Object.assign({}, this.data[id])`?

File: src/ObjectGraph.es6.js [hunk 4]
```diff
@@ -599,4 +622,4 @@ module.exports = facade(ObjectGraph, {
       this.prototype.blacklistedObjects.push(o);
     },
   },
-});
+});
```
Comment (mdittmer) [inline src/ObjectGraph.es6.js]: Please configure your editor to include a newline at the end of every file.

**Review by mdittmer — CHANGES_REQUESTED** [review]
Requested a few minor changes to consolidate metadata and avoid providing write access to raw data.

**Comment by jingt06** [comment]
fixed @mdittmer

**Review by mdittmer — CHANGES_REQUESTED** [review]
Preamble: My apologies for asking you to change-then-change-back on one point here.

I forgot that `metadata` is property-level, not object-level (i.e., `metadata[id][property-name] = property-metadata`). In light of this, we should probably store `functions` separately as you had before. Please keep the change with `Object.assign(...)` though.

**Comment by jingt06** [comment]
Moved functions out of metadata. @mdittmer

**Review by mdittmer — APPROVED** [review]
LGTM

**Pull request merged**